sdk/log/record.go

```diff
@@ -142,8 +181,103 @@ func (r *Record) AddAttributes(attrs ...log.KeyValue) {
 	r.back = append(r.back, attrs[i:]...)
 }
 
+func (r *Record) compactAttr() {
+	// index holds the location of attributes in the record based on the
+	// attribute key. If the value stored is < 0 the -(value + 1) (e.g. -1 ->
+	// 0, -2 -> 1, -3 -> 2) represents the index in r.nFront. Otherwise, the
+	// index is the exact index of r.back.
+	index := getIndex()
+	defer putIndex(index)
+
+	var dropped int
+	var cursor int
+	for i := 0; i < r.nFront; i++ {
+		key := r.front[i].Key
+		idx, found := index[key]
+		if found {
+			dropped++
+			r.front[-(idx + 1)] = r.front[i]
+		} else {
+			r.front[cursor] = r.front[i]
+			index[key] = -cursor - 1 // stored in front: negative index.
+			cursor++
+		}
+	}
+	r.nFront -= dropped
+
+	// Compact back storage into front.
+	for cursor < attributesInlineCount && len(r.back) > 0 {
+		key := r.back[0].Key
+		idx, found := index[key]
+		if found {
+			dropped++
+			r.front[-(idx + 1)] = r.back[0]
+		} else {
+			r.front[cursor] = r.back[0]
+			r.nFront++
+
+			index[key] = -cursor - 1 // stored in front: negative index.
+			cursor++
+		}
+		r.back = r.back[1:]
+	}
+
+	for i := 0; i < len(r.back); i++ {
+		key := r.back[i].Key
+		idx, found := index[key]
+		if found {
+			dropped++
+			if idx < 0 {
+				r.front[-(idx + 1)] = r.back[i]
+			} else {
+				r.back[idx] = r.back[i]
+			}
+			r.back = append(r.back[:i], r.back[i+1:]...)
+			i--
+		} else {
+			index[key] = i // stored in back: positive index.
+		}
+	}
+
+	r.dropped += dropped
+}
+
 // SetAttributes sets (and overrides) attributes to the log record.
 func (r *Record) SetAttributes(attrs ...log.KeyValue) {
+	// If adding these attributes could exceed limit, de-duplicate to minimize
```

Comment:
Are we required to de-duplicate to minimize overflow?  Duplicate attribute keys seems like something that should have undefined behavior.

We might be able to avoid repeatedly deduplicating if we did deduplication when the attributes are read, rather than in the write path.

Reply:
No, the existing attributes are going to be overwritten. The only de-duplication needed is on the input.

Reply:
Right. I'm saying that if we didn't have to deduplicate for overflow, we might be able to lazily deduplicate.  But it depends if we expect AddAttributes to be frequently called after SetAttributes

Reply:
Right. I wasn't sure if this was a requirement or not for duplicated attributes:

> attributes that should have been kept.

Reply:
I seem to remember this was a discussion we had prior here and in the specification SIG. I don't know if there was any documented outcome, but assume, given the trace implementation, it was decided we want accuracy over speed.

I'll try to look for more information about the previous conversation next week and post it here.

Reply:
I found https://github.com/open-telemetry/opentelemetry-go/issues/2554 which tracked our work on the span de-duplication. It mentions that [the specification states](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.8.0/specification/common/common.md):

> for each unique attributes key, addition of which would result in exceeding the limit, SDK MUST discard that key/value pair

Given the uniqueness requirement in the specification around limiting, this current implementation appears correct.